### PR TITLE
error-cause typedefs

### PR DESCRIPTION
**Pull request by conartist6**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by conartist6** [comment]
`Error: At 1:1 : Expected /\/\/ Type definitions for (non-npm package )?/`

Wow, very useful. I know just what to do.

**Comment by conartist6** [comment]
Though the error message obscures the real problem and I've followed every instruction to the letter, logic says that something is missing. I've never specified which version of the target package these are the typedefs for.

**Comment by conartist6** [comment]
Oh yes, because undocumented is the fact that `index.d.ts` must contain magic comments or else the parser just blows up and ejects its regex into the error log.

Whaddaya MEAN test the failure case!?!?! It worked fine for me!

File: types/error-cause/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Type definitions for error-cause 1.0
+// Project: https://github.com/es-shims/error-cause
+// Definitions by: Conrad Buck <https://github.com/conartist6>
```
Comment (ljharb) [inline types/error-cause/index.d.ts:3]:
```suggestion
// Definitions by: Conrad Buck <https://github.com/conartist6>
// Definitions by: Jordan Harband <https://github.com/ljharb>
```

File: types/error-cause/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+type ErrorTypes = [
+    'Error',
+    'AggregateError',
+    'EvalError',
+    'RangeError',
+    'ReferenceError',
+    'SyntaxError',
+    'TypeError',
+    'URIError',
+];
```
Comment (ljharb) [inline types/error-cause/index.d.ts:15]:
```suggestion
] as const;
```

**Review by ljharb — CHANGES_REQUESTED** [review]
This is missing types for AggregateError.

File: types/error-cause/index.d.ts (hunk 2 above)
Comment (conartist6) [reply types/error-cause/index.d.ts:15]: I don't think that compiles. It's a tuple type not an array literal. An array literal can't exist in a `d.ts` file.

File: types/error-cause/index.d.ts (hunk 1 above)
Comment (conartist6) [reply types/error-cause/index.d.ts:3]: I mean, really? I don't wanna be pedantic, but this was a lot of work and I'm the one who did it.

File: types/error-cause/index.d.ts (hunk 2 above)
Comment (ljharb) [reply types/error-cause/index.d.ts:15]: ohh right, you're correct. my bad

File: types/error-cause/index.d.ts (hunk 1 above)
Comment (ljharb) [reply types/error-cause/index.d.ts:3]: unfortunately it's the only way DefinitelyTyped will auto-tag me on future PRs to the types, or let me approve them on a review - it's not about claiming credit.

Comment (conartist6) [reply types/error-cause/index.d.ts:3]: I guess it's what should be there, as that way any future editors will get both of us as reviewers, which is what they should get.

**Review by ljharb — APPROVED** [review]
Thanks for all the work!

**Comment by conartist6** [comment]
A couple things.

Should the type of `cause` be `any` or `unknown`? I know it's more correct to use `unknown`, but typescript seems to err on the side of not getting in the way for patterns that it isn't ever really going to understand, which also makes sense to me.

Are you sure about `AggregateError`? Excluding it was an oversight, but on the other hand now that I look at it it's really nasty. For one thing, the whole set of typedefs is going to have to be TS4.3 for that to work because it seems that's the first version to include the `ES2021.promise` `lib` (unless you know a way around this).

Second typescript just doesn't like that pattern of inheritance. Yes in JS `Error` is in the proto chain, but from Typescript's perspective the constructor signatures are complete incompatible.

Finally lets go all the way back to the beginning. Should `AggregateError` have this functionality at all? I'd argue not -- it already has it! The cause of an aggregate error is whatever is in the `errors` array. The cause of the error is always no more and no less than every single one of those errors. That change of course would go all the way back to the way spec and implementation. Thoughts?

**Comment by ljharb** [comment]
I would strongly prefer to use `unknown`, and never use `any`, as that defeats the entire point of using a type system.

I think it would be fine to land this without AggregateError for now, but we'll want to have dual typedefs so that in TS4.3+, it uses those types. Also, see the types for `es-aggregate-error`, where those types may already be defined.

The spec itself includes the cause on AggregateError, so that ship has sailed.

**Comment by conartist6** [comment]
Oof. Well that would mean we could ship what's here right now.

I also wouldn't mind just leaving `AggregateError` out of the typedefs for the long run. JS is always going to permit some things that TS thinks are errors, which is fine.

**Comment by ljharb** [comment]
It definitely can't be left out; it's part of the API, and the spec. We can ship this now tho and iterate afterwards, of course.

**Comment by conartist6** [comment]
Guess it'll get a lint rule then, and hopefully the standard linter bases will pick it up.

**Comment by conartist6** [comment]
Rebuilt to fix failed parsing of multiple author lines.

Any ideas on how to start emitting this information in a sane way? The polyfill is really the easy part. The hard part is the most VMs are just going to ignore that information. It needs to be captured at the moment that error objects are stringified. I'm considering forking the code that jest uses to print its errors. [pretty-error](https://github.com/AriaMinaei/pretty-error) has existing docs on how to attach custom error rendering code in a variety of real life scenarios...

**Review by ljharb — APPROVED** [review]

**Comment by conartist6** [comment]
Thanks! And thanks for your time. I can't wait to start going in and actually fixing these problems so that all the information necessary to debug an error is present where it should be.

And since I have your ear LMK if you know of any companies looking for people who can be a bit obsessive about developer experience. I'm currently working on [iter-tools](https://github.com/iter-tools/iter-tools) and macrome, a build system for effortless management of source-colocated checked-in assets.

**Review by peterblazejewicz — APPROVED** [review]
@conartist6 thx!
LGTM!

**Comment by conartist6** [comment]
Ready to merge

**Pull request merged**